src/mgr/TTLCache.h

```diff
@@ -0,0 +1,159 @@
+      content.insert({key, value});
+    }
+  }
+  Value get(Key key) { return content[key]; }
+  void erase(Key key) { content.erase(content.find(key)); }
+  void clear() { content.clear(); }
+  bool exists(Key key) { return content.find(key) != content.end(); }
+  std::pair<uint64_t, uint64_t> get_hit_miss_ratio() {
+    return std::make_pair(hits.load(), misses.load());
+  }
+  int size() { return content.size(); }
+  ~Cache(){};
+};
+
+typedef std::chrono::time_point<std::chrono::steady_clock> ttl_time_point;
+template <class Key, class Value>
+class TTLCache : public Cache<Key, std::pair<Value, ttl_time_point>> {
+private:
+  uint16_t ttl;
+  typedef std::pair<Value, ttl_time_point> value_type;
+  float ttl_spread_percent;
+
+  Value get_value(Key key);
+  ttl_time_point get_value_time_point(Key key);
+  bool exists(Key key);
+  bool expired(Key key);
+  void finish_get(Key key);
+  void finish_erase(Key key);
+
+public:
+  TTLCache(uint16_t ttl_ = 0, uint16_t size = UINT16_MAX, float spread = 0.25)
+      : Cache<Key, value_type>(size), ttl{ttl_}, ttl_spread_percent{spread} {}
+  ~TTLCache(){};
+  void insert(Key key, Value value);
+  boost::optional<Value> get(Key key);
+  void erase(Key key);
+  void clear();
+  void set_ttl(uint16_t ttl);
+};
+
+template <class Key, class Value>
+void TTLCache<Key, Value>::insert(Key key, Value value) {
+  auto now = std::chrono::steady_clock::now();
+
+  if (!ttl)
+    return;
+  uint16_t ttl_spread = ttl * this->ttl_spread_percent;
+  // in order not to have spikes of misses we increase or decrease by 25% of
+  // the ttl
+  uint16_t spreaded_ttl = ttl + (rand() % (ttl_spread * 2) + (-ttl_spread));
+  auto expiration_date = now + std::chrono::seconds(spreaded_ttl);
+  this->Cache<Key, value_type>::miss();
+  this->Cache<Key, value_type>::insert(key, {value, expiration_date});
+}
+
+template <class Key, class Value>
+boost::optional<Value> TTLCache<Key, Value>::get(Key key) {
+  boost::optional<Value> value;
+  if (!this->exists(key)) {
+    return value;
+  }
+  if (expired(key)) {
+    erase(key);
+    return value;
+  }
+  this->Cache<Key, value_type>::hit();
+  finish_get(key);
+
+  value = {get_value(key)};
+  return value;
+}
+
+template <class Key, class Value> void TTLCache<Key, Value>::erase(Key key) {
+  finish_erase(key);
+  this->Cache<Key, value_type>::erase(key);
+}
+
+template <class Key, class Value> bool TTLCache<Key, Value>::expired(Key key) {
+  ttl_time_point expiration_date = get_value_time_point(key);
+  auto now = std::chrono::steady_clock::now();
+  if (now >= expiration_date) {
+    return true;
+  } else {
+    return false;
+  }
+}
+
+template <class Key, class Value> void TTLCache<Key, Value>::clear() {
+  this->Cache<Key, value_type>::clear();
+}
+
+template <class Key, class Value>
+Value TTLCache<Key, Value>::get_value(Key key) {
+  value_type stored_value = this->Cache<Key, value_type>::get(key);
+  Value value = std::get<0>(stored_value);
+  return value;
+}
```

Comment:
I think this one is needed since we usually only want to access the real value instead of the whole pair which contains de time point. Furthermore, it is used in different places.

Reply:
Ok, makes sense!